### PR TITLE
DOP-336 : add gcs publisher

### PR DESCRIPTION
**Pull request by ashamshad**
https://liveramp.atlassian.net/browse/DOP-336

This is to add a Publisher able to publish packages to GCS bucket.

Changes :
Added a new GCSPublisher who's target is a GCS bucket.
Adapted the creation of the Publisher into a Factory. Unit tests added for the Factory

To do :
Tests still need to be added on the GCSPublisher, will mock and inject the GoogleCloudStorageClient for that.

### CODE REVIEW

File: packager/src/main/java/io/cdap/hub/GoogleCloudStorageClient.java
Comment (sagarrbatchu) [inline packager/src/main/java/io/cdap/hub/GoogleCloudStorageClient.java:23]: for mocking behavior in tests i would also glance at how google does it for their GCS plugins in https://github.com/data-integrations/google-cloud

File: packager/src/main/java/io/cdap/hub/publisher/GCSPublisher.java
Comment (sagarrbatchu) [inline packager/src/main/java/io/cdap/hub/publisher/GCSPublisher.java:130]: comment still references `s3`

File: packager/src/main/java/io/cdap/hub/publisher/GCSPublisher.java
Comment (sagarrbatchu) [inline packager/src/main/java/io/cdap/hub/publisher/GCSPublisher.java:116]: `LOG.info` still references `s3`

File: packager/src/main/java/io/cdap/hub/publisher/GCSPublisher.java
Comment (sagarrbatchu) [inline packager/src/main/java/io/cdap/hub/publisher/GCSPublisher.java:161]: Comment still references `s3`

File: packager/src/main/java/io/cdap/hub/publisher/GCSPublisher.java
Comment (sagarrbatchu) [inline packager/src/main/java/io/cdap/hub/publisher/GCSPublisher.java:191]: Comment still references s3

File: packager/pom.xml
Comment (sagarrbatchu) [inline packager/pom.xml:62]: just curious about this android dependency :)

**Review by sagarrbatchu — COMMENTED** [review]
Looks good so far ! 

I don't know if i caught all of them but would do a `ctrl-f` for "s3" in the `GCS*` files and replace with gcs

File: packager/pom.xml
Comment (ashamshad) [reply packager/pom.xml:62]: https://mvnrepository.com/artifact/com.google.guava/guava after the 23.0, every dependency is either `-jre` or `-android`
I used the `-android` version to match the one used by the google cloud storage dependency

File: packager/src/main/java/io/cdap/hub/Tool.java
Comment (BaptBeuns) [inline packager/src/main/java/io/cdap/hub/Tool.java:24]: Since this is abstracted, I feel like that import is not needed?

Comment (BaptBeuns) [reply packager/src/main/java/io/cdap/hub/Tool.java:24]: Oh I see that `Tool: getPublisher` returns that type. Should it be changed?

File: packager/src/main/java/io/cdap/hub/publisher/PublisherFactory.java
Comment (BaptBeuns) [inline packager/src/main/java/io/cdap/hub/publisher/PublisherFactory.java:30]: Good job on creating a factory

File: packager/src/main/java/io/cdap/hub/publisher/GCSPublisher.java
Comment (BaptBeuns) [inline packager/src/main/java/io/cdap/hub/publisher/GCSPublisher.java:51]: Would this class benefit from unit tests at all?

**Review by BaptBeuns — APPROVED** [review]

File: packager/src/main/java/io/cdap/hub/publisher/GCSPublisher.java
Comment (ashamshad) [reply packager/src/main/java/io/cdap/hub/publisher/GCSPublisher.java:51]: yes unit tests for this class are in progress, just trying the right way to mock the Storage

File: packager/src/main/java/io/cdap/hub/Tool.java
Comment (ashamshad) [reply packager/src/main/java/io/cdap/hub/Tool.java:24]: yes forgot to remove that unused method, good spot

File: packager/src/main/java/io/cdap/hub/publisher/GCSPublisher.java
Comment (BaptBeuns) [reply packager/src/main/java/io/cdap/hub/publisher/GCSPublisher.java:51]: I think someone from Ingestion made a utility class for that. Let me look for it.

File: packager/src/main/java/io/cdap/hub/GoogleCloudStorageClient.java
Comment (ashamshad) [reply packager/src/main/java/io/cdap/hub/GoogleCloudStorageClient.java:23]: package limited not sure, as Publisher related classes are in publisher package, and this one doesnt fit to go in publisher package I think. 

if I make `createStorageConnection` a `static` method, then I can't mock it.

File: packager/pom.xml
Comment (ashamshad) [reply packager/pom.xml:62]: The google cloud storage dependency uses the android one, so I just matched it. 
I can investigate about jre and android versions though

File: packager/src/main/java/io/cdap/hub/publisher/GCSPublisher.java
Comment (BaptBeuns) [reply packager/src/main/java/io/cdap/hub/publisher/GCSPublisher.java:51]: I found this old thread where they talk about it. You might have to look for the right code version to find the appropriate code. The important part apparently is:
```
  private static StorageOptions storageOptions = LocalStorageHelper.getOptions();
```

https://liveramp.slack.com/archives/CCE236DFT/p1573591855066600

File: packager/src/main/java/io/cdap/hub/GoogleCloudStorageClient.java
Comment (ashamshad) [reply packager/src/main/java/io/cdap/hub/GoogleCloudStorageClient.java:23]: https://github.com/data-integrations/google-cloud/blob/develop/src/test/java/io/cdap/plugin/gcp/gcs/StorageClientTest.java 
Looked at this test class, there's nothing about mocking the actual Storage. Nothing either in the other plugins about mocking out the storage.

There's this one apparently that we can use : https://github.com/googleapis/java-storage-nio/blob/master/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/testing/LocalStorageHelper.java
Will try that out otherwise will just mock the required behaviour with

File: packager/src/main/java/io/cdap/hub/publisher/PublisherFactory.java
Comment (sagarrbatchu) [reply packager/src/main/java/io/cdap/hub/publisher/PublisherFactory.java:30]: + 1

**Review by sagarrbatchu — APPROVED** [review]

**Review by BaptBeuns — CHANGES_REQUESTED** [review]
Looks good and I like the tests. I wonder now what the code coverage is for this change :D 

I have a couple of comments on some classes/attributes that are not needed, but the rest LGTM.

Have you installed LiveRamp's style guide to your Intellij? Some classes are not consistent with that style guide. I'll approve when the style adheres to our guide.

File: packager/src/main/java/io/cdap/hub/GoogleCloudStorageClient.java
Comment (ashamshad) [reply packager/src/main/java/io/cdap/hub/GoogleCloudStorageClient.java:23]: GoogleCloudStorageClient is indeed used, I use it in the unit test to inject a fake storage, for example have a look at GCSPublisher:62, and how the GCSPublisher is created with `GCSPublisher.builder(someProjectId, "someBucket")
                .buildWithStorageClient(googleCloudStorageClient)`

File: categories.json
Comment (ashamshad) [reply categories.json:1]: yes this file is not supposed to be pushed, it was pushed by accident I think as it's in the .gitignore too.
this file and the packages.json file are generated each time we run the hub build step

File: packager/src/test/java/io/cdap/hub/CommandLineHelperTest.java
Comment (ashamshad) [reply packager/src/test/java/io/cdap/hub/CommandLineHelperTest.java:20]: this is Junit 5, it has switched to junit.jupiter now

File: packager/src/main/java/io/cdap/hub/publisher/GCSPublisher.java
Comment (ashamshad) [reply packager/src/main/java/io/cdap/hub/publisher/GCSPublisher.java:56]: This object is stubbed and injected by the unit test, so we need this client layer on top of the storage for the testing.
Also the Storage object is initialized in the `publish` method through this client object, and kept in memory just for the sake of not creating a connection for each GCS operation.

I also think that in general it's better to encapsulate the client connection creation in a separate object, so it's easier to mock and it's isolated from the actual code

File: packager/src/main/java/io/cdap/hub/publisher/GCSPublisher.java
Comment (ashamshad) [reply packager/src/main/java/io/cdap/hub/publisher/GCSPublisher.java:131]: yes indeed, this method is private and only called by the method `publish`

File: packager/pom.xml
Comment (ashamshad) [reply packager/pom.xml:135]: jupiter is just the new name for JUnit 5. The migration to 5 is quite easy because they've kept most of the API similar

**Comment by ashamshad** [comment]
> Looks good and I like the tests. I wonder now what the code coverage is for this change :D
> 
> I have a couple of comments on some classes/attributes that are not needed, but the rest LGTM.
> 
> Have you installed LiveRamp's style guide to your Intellij? Some classes are not consistent with that style guide. I'll approve when the style adheres to our guide.

Installed the Liveramp style guide yes, thanks

**Comment by ashamshad** [comment]
@BaptBeuns alright so I've made some changes following your remarks on the `googleCloudStorageClient` field parameter.
I've kept only constructor now, which takes the client as parameter, so removed the other one.
The `builder()` method takes the `googleCloudStorageClient` as parameter so it's now enforced.
The `Storage` connection init is done now in the constructor 
By cleaning all that, I also managed to remove the `projectId` as field parameter from GCSPublisher !

So I think now we've got a cleaner version, please have another look

File: packager/src/main/java/io/cdap/hub/GoogleCloudStorageClient.java
Comment (sagarrbatchu) [reply packager/src/main/java/io/cdap/hub/GoogleCloudStorageClient.java:23]: sounds good

**Review by sagarrbatchu — APPROVED** [review]
Looks good to me !

**Review by BaptBeuns — APPROVED** [review]

File: packager/src/main/java/io/cdap/hub/CommandLineHelper.java
Comment (BaptBeuns) [inline packager/src/main/java/io/cdap/hub/CommandLineHelper.java:25]: note: apply style guide

File: packager/src/test/java/io/cdap/hub/CommandLineHelperTest.java
Comment (BaptBeuns) [inline packager/src/test/java/io/cdap/hub/CommandLineHelperTest.java:22]: note: I like avoiding to use `import static`

**Pull request merged**
